docs/source/research.rst

```diff
@@ -0,0 +1,547 @@
+--------------------------------------------
+*Slim yet handsome validation library.*
+
+.. code-block:: python
+
+    from good import Schema, Entire
+
+    def maxkeys(n):
+        # Return a validator function
+        def validator(d):
+            # `d` is the dictionary.
+            # Validate it
+            assert len(d) <= 3, 'Dict size should be <= 3'
+            # Return the value since all callable schemas should do that
+            return d
+        return validator
+
+    schema = Schema({
+        str: int,
+        Entire: maxkeys(3)
+    })
+
+
+`incoming <https://incoming.readthedocs.org/>`_
+-----------------------------------------------
+*JSON validation framework for Python.*
+
+.. code-block:: python
+
+    >>> class AddressValidator(PayloadValidator):
+    ...     street = datatypes.String()
+    ...     country = datatypes.String()
+    ...
+    >>> class PersonValidator(PayloadValidator):
+    ...     name = datatypes.String()
+    ...     age = datatypes.Integer()
+    ...     address = datatypes.JSON(AddressValidator)
+    ...
+    >>> PersonValidator().validate(dict(name='Some name', age=19, address=dict(street='Brannan, SF', country='USA')))
+    (True, None)
+    >>>
+    >>> PersonValidator().validate(dict(name='Some name', age=19, address=dict(street='Brannan, SF', country=0)))
+    (False, {'address': ['Invalid data. Expected JSON.', {'country': ['Invalid data. Expected a string.']}]})
+
+
+`Kanone <https://github.com/doncatnip/kanone>`_
+-----------------------------------------------
+*A general purpose validation library*
+
+.. code-block:: python
+
+    >>> from kanone import *
+    >>> HelloSchema = Schema\
+        ( 'nick'
+            , String() & Len(max=20)
+        , 'email'
+            , web.Email()
+        , 'email_confirm'
+            , Match( Field('.email'), ignoreCase=True )
+        )
+    >>> context = HelloSchema.context\
+        (   { 'nick':'bob'
+            , 'email':'Bob@Some.Domain.Org'
+            , 'email_confirm': 'BOB@Some.domain.org'
+            }
+        )
+    >>> context('nick').result
+    u'bob'
+    >>> context('email').result
+    u'Bob@some.domain.org'
+
+
+`lasso <https://lasso.readthedocs.org/en/latest/>`_
+---------------------------------------------------
+*Lightweight module to define serializable, schema-validated classes*
+
+.. code-block:: python
+
+    >>> class Name(lasso.Schemed):
+    ...     __schema__ = { "first": str, "family": str }
+    ...
+    >>> class User(lasso.Schemed):
+    ...     __schema__ = { "name": Name, "email": str }
+    ...
+    >>> jdoe = User(name=Name(first="John", family="Doe"),
+    ...             email="j@doe.org")
+    ...
+
+
+`marshmallow: simplified object serialization <https://marshmallow.readthedocs.org/>`_
+--------------------------------------------------------------------------------------
+*marshmallow is an ORM/ODM/framework-agnostic library for converting complex datatypes, such as objects, to and from
```

Comment:
At least github doesn't render this line correctly, maybe rst doesn't support multiline `*foo*`?